src/coreclr/src/jit/hwintrinsic.cpp

```diff
@@ -701,22 +730,40 @@ GenTree* Compiler::impHWIntrinsic(NamedIntrinsic        intrinsic,
                         op1 = op1->gtGetOp1();
                     }
                 }
-                retNode = gtNewSimdHWIntrinsicNode(retType, op1, intrinsic, baseType, simdSize);
+
+                retNode = isScalar ? gtNewScalarHWIntrinsicNode(retType, op1, intrinsic)
+                                   : gtNewSimdHWIntrinsicNode(retType, op1, intrinsic, baseType, simdSize);
                 break;
             }
 
             case 2:
             {
-                argType = JITtype2varType(
-                    strip(info.compCompHnd->getArgType(sig, info.compCompHnd->getArgNext(argList), &argClass)));
+                CORINFO_ARG_LIST_HANDLE arg2 = info.compCompHnd->getArgNext(argList);
+                argType = JITtype2varType(strip(info.compCompHnd->getArgType(sig, arg2, &argClass)));
                 op2 = getArgForHWIntrinsic(argType, argClass);
 
                 op2 = addRangeCheckIfNeeded(intrinsic, op2, mustExpand);
 
                 argType = JITtype2varType(strip(info.compCompHnd->getArgType(sig, argList, &argClass)));
                 op1     = getArgForHWIntrinsic(argType, argClass);
 
-                retNode = gtNewSimdHWIntrinsicNode(retType, op1, op2, intrinsic, baseType, simdSize);
+                retNode = isScalar ? gtNewScalarHWIntrinsicNode(retType, op1, op2, intrinsic)
+                                   : gtNewSimdHWIntrinsicNode(retType, op1, op2, intrinsic, baseType, simdSize);
+#ifdef TARGET_XARCH
+                if (intrinsic == NI_SSE42_Crc32 || intrinsic == NI_SSE42_X64_Crc32)
+#endif
+#ifdef TARGET_ARMARCH
+                if (intrinsic == NI_Crc32_ComputeCrc32 || intrinsic == NI_Crc32_ComputeCrc32C ||
+                    intrinsic == NI_Crc32_Arm64_ComputeCrc32 || intrinsic == NI_Crc32_Arm64_ComputeCrc32C)
+#endif
+                {
+                    // type of the second argument
+                    CorInfoType corType = strip(info.compCompHnd->getArgType(sig, arg2, &argClass));
+
+                    // TODO - currently we use the BaseType to bring the type of the second argument
```

Comment:
The `TODO` comments should be more specific - I think this is probably a `TODO-Cleanup`?

Reply:
I just copied the existing TODO comment along with the code.